aptos-move/move-examples/token_objects/marketplace/sources/marketplace.move

```diff
@@ -0,0 +1,1809 @@
+    use std::option::{Self, Option};
+    use std::signer;
+    use std::string::{Self, String};
+    use std::vector;
+
+    // -- Errors --
+
+    /// Listing not found
+    const ELISTING_NOT_FOUND: u64 = 1;
+    /// Price below minimum listing price
+    const EPRICE_BELOW_MINIMUM: u64 = 2;
+    /// Listing has expired
+    const ELISTING_EXPIRED: u64 = 3;
+    /// Cannot remove listing, signer is not seller of token
+    const ENOT_SELLER: u64 = 4;
+    /// Cannot list token, signer is not owner of token
+    const ENOT_OWNER: u64 = 5;
+    /// Listing expiration time is less than 5 minutes
+    const ETOO_SHORT_DURATION: u64 = 6;
+    /// Denominator cannot be 0
+    const EDENOMINATOR_ZERO: u64 = 7;
+    /// Commission is 100% or greater
+    const ECOMMISSION_TOO_HIGH: u64 = 8;
+    /// Marketplace for coin not found
+    const EMARKETPLACE_NOT_FOUND: u64 = 9;
+    /// Not authorized to modify marketplace, not the owner of the marketplace
+    const ENOT_AUTHORIZED: u64 = 10;
+    /// Bid must be strictly larger than the previous bid
+    const EBID_TOO_LOW: u64 = 11;
+    /// No bid found
+    const ENO_BID_FOUND: u64 = 12;
+    /// Auction not yet over
+    const EAUCTION_NOT_OVER: u64 = 13;
+    /// Listing not on Marketplace
+    const ELISTING_NOT_ON_MARKETPLACE: u64 = 14;
+    /// Collection bid amount cannot be zero
+    const ECOLLECTION_BID_AMOUNT_ZERO: u64 = 15;
+
+    // -- Constants --
+
+    const FIVE_MINUTES_SECS: u64 = 300;
+
+    // These delineate supported object types for listing
+    const TOKEN_V1: vector<u8> = b"TokenV1";
+    const TOKEN_V2: vector<u8> = b"TokenV2";
+
+    // -- Structs --
+
+    /// A standalone resource that represents the marketplace in it's entirety
+    ///
+    /// This also allows you to list in other currencies than the native coin
+    struct Marketplace<phantom CoinType> has key {
+        fee_schedule: FeeSchedule,
+        extend_ref: ExtendRef,
+        /// An event stream of changes to the fee schedule
+        fee_mutation_events: event::EventHandle<FeeMutationEvent>,
+    }
+
+    /// The desciption of all fees on the marketplace, it allows for
+    /// listing fees to prevent spam, as well as commissions to take a percentage
+    /// of each sale.
+    struct FeeSchedule has store, drop {
+        /// Address to send fees to
+        fee_address: address,
+        /// A fixed fee for making a listing
+        listing_fee: u64,
+        /// A fixed fee for making a bid
+        bid_fee: u64,
+        /// Numerator on the percentage fee of a listing sale
+        commission_numerator: u64,
+        /// Denominator on the percentage fee of a listing sale
+        commission_denominator: u64,
+    }
+
+    /// A listing for a single Token object on the market place
+    ///
+    /// The listing will own the token until it is bought by a buyer or it is
+    /// removed from listing.
+    /// TODO: Allow extending the expiration time?
+    struct Listing<phantom CoinType> has key {
```

Comment:
also should have a pointer back to the marketplace that owns this, since that's the one that ultimately gets the majority of the fees... other marketplace aggregators can wrap around this, but I think that's another concept.